# The nature of data

## Basic data types in Python

Before we begin our Python tutorial, here's one important note about code cells in Jupyter notebooks: If you have a code cell and only one result generates output, we don't need to use a `print` function.

In [1]:
a = 1
a # this line is the reason the cell prints 1 in the output

1

But if we have more than one line that should generate output, we need to use `print` otherwise only the **last** line will produce output.

In [2]:
a, b = 1, 2
a # will not produce output
b # will produce output

2

In [3]:
a, b = 1, 2
print(a) # will produce output
print(b) # will produce output

1
2


If you want to produce output for something that is running inside of a loop, you also need to use `print`.

In [4]:
for i in range(5):
    print(i)

0
1
2
3
4


Let's begin by learning about the basic building blocks of Python data structures: **integers, floats, booleans, and strings**.

In [5]:
30 + 10

40

In [6]:
type(40)

int

In [7]:
4.5

4.5

In [8]:
type(40.0)

float

In [9]:
int(40.0)

40

In [10]:
#Means not a number
float('nan')

nan

In [11]:
#inifitiy float value
float('inf')

inf

In [12]:
float('test')

ValueError: could not convert string to float: 'test'

### Exercise (1 minute)

Verify that `float('inf')` indeed behaves like infinity. Mathematically, what should happen if you multiply infinty by 2? What about dividing infinity by itself?

In [ ]:
float('inf') / float('inf')


In [ ]:
float('inf') * 2

In [ ]:
float('inf') + 10

### End of exercise

In [ ]:
1 == 0

In [ ]:
1 is 1

In [ ]:
1 in [1, 0]

In [ ]:
not(1 == 0)

In [ ]:
type(1 == 0)

What sets dynamic programming languages like Python apart from **strongly-typed** programming languages like Java is that we can change the type of a variable at execution time. In strongly-typed programming languages we define the type ahead of time and enforce it at runtime. In Python we can do things like this:

In [ ]:
1.0 + False

In [ ]:
1 + 1.0

In [ ]:
3 - True # this is the kind of thing that you shouldn't do, even if you can, True gets treated as 1

Using these constructs, we can sometimes write conditional statements in a concise way. But be careful doing this, because as popular as **one-liners** are in Python, it's important to write code that is also easy to understand, either to you six months from now, or to someone else reading your code. Here's an example:

In [ ]:
statement = (2 > 4)
#False is treated as 0
print(statement * 5 + (1 - statement)*3) # hard to understand

if statement == True: # easier to understand but many lines
    print(5)
else:
    print(3)

if statement: # turns out we don't need the `== True` part, it's implied
    print(5)
else:
    print(3)

if not statement: # turns out we don't need the `== True` part, it's implied
    print(3)
else:
    print(5)

print(5 if statement else 3) # both concise and easy to understand, but a little risky

Of course this is both a convenience and an inconvenience. It is a convenience because it means that we can write very short code to do what we need to do. In data science, where we want to quickly analyze our data, create summaries and visualizations, train and test models, etc. this convenience really pays off. However, the inconvenience lies in that we are more susceptible to have bugs in our program that we do not catch until later. So we need to be careful and develop best practices when we code in Python. 

In [ ]:
s = "abc"

This problem can be amplified inside a Jupyter notebook where we are not always careful and might execute a cell twice by accident, such as in the cell below. To avoid this kind of problem, we should combine the above and below cells into one.

In [ ]:
s + "def"

In a way we can say that dynamic languages **try to be smart when we are being lazy** by guessing what we mean when we type some code. Of course there is a limit to that and sometimes we need to be a little more explicit, such as here when we need to coerce a string into an integer or vice versa for the operation to make sense.

In [ ]:
int('1') + 2

In [ ]:
str(1) + '2'

From these basic building blocks, we create the next set of fundamental objects: **lists, tuples, sets and dictionaries**.

In [ ]:
fruits = ["apple", "pear", "pitaya", "zapote"] # lists
print(fruits[0]) # first element
print(fruits[-1]) # last element
print(fruits[0:2]) # first two elements

Lists are **mutable** objects, meaning that we can change its elements, add new ones, or drop elements from it.

In [ ]:
fruits[-1] = "nispero" # lists are mutable

In [ ]:
fruits + ["guanabana", "papaya"] # adding something to a list is easy

In [ ]:
#Since youre not using fruits = fruits +['papaya'] then it does not stick
fruits + ["papaya"]

In [ ]:
del fruits[:2] # dropping something from a list is easy
print(fruits)

We can have lists of booleans, strings, or even lists of mixed data types. Basically, lists are very **flexible** objects, and we use them when we need flexibility. Just know that there's no such thing as a free lunch: when you have more flexibility, you usually have less efficiency.

Notice what happens when we convert a string into a `list`:

In [ ]:
list("abc")

We can join things back using `join`:

- will be used heavily during text analysis

In [ ]:
"-".join(list("abc")) # we can use join to join a list of strings

With a list of booleans, we can use `any` and `all` to check conditions. We later encounter similar versions of these functions that work on a `DataFrame`. They have the same name but slightly different functionality.

In [ ]:
any([True, False, False])

In [ ]:
all([True, False, False])

Finally, lists can be **nested** too, meaning we can have lists of lists.

In [ ]:
nested_list = [[-1, 1, 9], [-2, 3, 4]]
print(nested_list)

In [ ]:
nested_list[1][0]

There is a really neat way to create lists with Python using what's called **list comprehension**. It's really just a shortcut for creating a list very quickly and in a way that makes the code look easy to follow.

In [ ]:
some_list = [] # initialize empty list
for i in range(10):
    if i % 3 == 1: # if i divided by 3 leaves a remainder of 1
        some_list += [i] # add i to the list

print(some_list)

The above snippet is valid, but there's a much easier way of doing it using a list comprehension:

 - preferred way for a for loop is below, smoother and cleaner


In [ ]:
some_list = [i for i in range(10) if i % 3 == 1]
print(some_list)

Tuples look somewhat similar to lists, but they are much more rigid. Tuples are **immutable** meaning that once they're created they can't be changed (unless you recreate them).

In [ ]:
tup = tuple([2, 4, 12, "bla"])
tup = (2, 4, 12, "bla")
print(tup)

### Exercise (1 minute)

Try changing the tuple above by changing one of its elements and see what happens. What happens when trying to add tuples to each other? Compare that with lists.

In [ ]:
tup[1] = 1

In [ ]:
tup2 = tuple([8,9,10])

In [ ]:
tup3 = tup+tup2
tup3

### End of exercise

Tuples are handy when you know what information goes in each place and don't need the flexibility to change it later.

We can also have **sets** in Python, which are similar to mathematical sets: elements in a set are unique and the order doesn't matter. Sets have certain **methods** that sound familiar if you remember sets in math, such as `difference`.

- no duplicates in sets, all the characters are unique

In [ ]:
my_set = set([2, 2, "hello", "hello", 5])
my_set = {2, 2, "hello", "hello", 5}
print(my_set)

In [ ]:
my_set.difference(set([2, 4, "hello"]))

If you're not familiar with the term **method** it refers to the fact that once you create an object, there are certain functions that you can call that are relevant to the object. For example, a set has a method called `difference` so that you can take the difference of one set from another set. To call the method, you type the name of the object itself, followed by a period, and the name of the method. Tab completion can help you find out what all methods an object might have. Similarly, on more sophisticated editors like Visual Studio Code, we can use intellisense to quickly view the methods and attributes of an object and even pull up their definitions. Finally, we can also call the `dir` function in Python for this. In the below example, notice how we drop methods that start with a double underscore (called "dunder" by Pythonistas) as these are the internal methods and we should avoid using them. A deeper discussion of this is out of the scope of this lesson.

In [ ]:
[s for s in dir(my_set) if s[:2] != '__']

### Exercise (2 minutes)

Explore the methods and attributes associated with a list and a tuple. Just create a list, then type its name, followed by dot, and press `SHIFT+TAB` to let auto-complete show you the methods. Then try the same thing for a tuple. Can you write a code to see which methods  or attributes (if any) the two objects have in common?

### End of exercise

A Python dictionary is similar to a list, but instead of elements being ordered by their position in the list, the elements are **values** that are paired with **keys**, which makes dictionaries a **key-value pair**.

In [ ]:
my_dict = {"a": 12, "b": "hello", "c": [4, 9], "d": {"one": 1, "two": 2}}
print(my_dict)

In [ ]:
my_dict.keys()

In [ ]:
my_dict.values()

In [ ]:
del my_dict['a']
print(my_dict)

There is one more data type we briefly cover here, although it is not a **native** data type (a.k.a **built-in**), in the sense that we need to load a library for it: the `numpy` library. The `numpy` library is a library used to do linear algebra calculations. The most basic `numpy` object is an `array`.

In [ ]:
import numpy as np
arr = np.array([2, 4, 1, 9])
print(arr)

In [ ]:
arr.mean()

In [ ]:
np.mean(arr)

The above array looks like a Python list, but a `numpy` array and a list are very different. Let's look at some example to see what makes them different.

### Exercise (2 minutes)

In the following cell, we create a Python list and a `numpy` array. Then we perform the same operation on the list and on the array, but get very different results. Explain what happens in each case.

In [ ]:
lst = [2, 4, 9, 7, 1, 9] # this is a list
arr = np.array(lst) # this is an array

print(lst + [1])
print(arr + [1])

Return to the above code and change one of the elements of the list from an integer to a string, then report what happens.

### End of exercise

As we saw in the last exercise, `numpy` arrays cannot contain elements of different types. We say that they are `atomic`. This makes `numpy` arrays ideal for storing columns in tables, because in a table, different columns can be of different types, but for a given column all elements (rows) are expected to be of the same type.

The `numpy` library is meant to make it easy to work with arrays. We saw for example how we can add 1 to every element of the array **without having to write a loop or list comprehension**. Let's look at another example: say we want to filter an array to only keep elements greater than 5. Compare the list implementation and the array implementation:

In [ ]:
lst = [2, 4, 9, 7, 1, 9] # this is a list

[i for i in lst if i > 5] # filtering a list

In [ ]:
arr = np.array(lst) # this is an array

arr[arr > 5] # filtering an array

There's a lot more we could be talking, but this is a data science class, not a class on Python programming. We **strongly** encourage you to take an introductory Python class and get very comfortable with the constructs we covered here. 

## Other useful functions

For the remainder of the notebook, we cover a few more things that have relevance to data science and then try to bring these examples home by looking at a practical example.

We often have to print some summaries about data. The `print` function along with the `format` method can be very helpful.

In [13]:
dec, bignum, pct = 3.1415, 4.234**5, 0.89 # you can assign multiple variable at once with this shortcut
print("The three numbers are {}, {}, and {}.".format(dec, bignum, pct))
print("The decimal rounded to 2 digits is {:0.3}".format(dec))
print("The bignum with thousand separators and 2 decimals is {:,.2f}".format(bignum))
print("The percentage number is {:2.0f}%".format(pct * 100))

The three numbers are 3.1415, 1360.6745706140914, and 0.89.
The decimal rounded to 2 digits is 3.14
The bignum with thousand separators and 2 decimals is 1,360.67
The percentage number is 89%


Sometimes we want to run some code, and in case it fails, run some other code. This can help us **gracefully** (that's a technical term) handle errors in our code. To do that, we use the `try` and `except` function.

In [14]:
try:
    new_var += 1 # new_var doesn't exist
    print("Found and incremented new_var. The new value is {}.".format(new_var))
except:
    new_var = 45
    print("Initialized new_var to 45")

Initialized new_var to 45


Try running the above cell multiple times to see what happens. The `try` function can be very helpful when we're traversing data that is not well structured and we expect to run into errors but don't want the errors to stop us mid-stream.

Writing functions in Python is relatively easy:

In [15]:
def my_function(n, m = 1): # name and arguments for the function
    return(n + m)

print(my_function(3))
print(my_function(n = 3))

4
4


In [16]:
print(my_function(3, 4)) # you can match arguments by position, here order matters
print(my_function(n = 3, m = 4)) # you can match arguments by name
print(my_function(m = 4, n = 3)) # if you match arguments by name, order doesn't matter

7
7
7


### Exercise (3 minutes)

In the list below, one of the elements is a string by accident. Write a program that multiplies each element of the list by itself. Use `try` and `except` to leave the element as-is when the element is not a number.

There are different ways to solve this. Here are two ways we propose:
- use a loop that iterates over the index of the list using `range(len(my_list))`
- write a function and then use list comprehension to apply it to each element of the list

In [17]:
my_list = [2, 4, 8, "3", 5]

def convert(n):
    ## write code here
    return n

[convert(i) for i in my_list]

[2, 4, 8, '3', 5]

### End of exercise

One final note about functions. The above function was created in the current Python session, so we can call it in the same session. But what about calling a function from one of the libraries we load. For example, let's say we want to use the `mean` function in the `numpy` library. We already saw that we first have to load `numpy` and then call `numpy.mean` or `np.mean` if we alias numpy with `np`.

In [18]:
import numpy as np
np.mean([4, 8, 3])

5.0

However, if we don't like to preface the function name with the library name, we can load the function like this:

In [19]:
from numpy import mean
mean([4, 8, 3])

5.0

However, we should be careful about doing this because we might overwrite an existing function with the same name. This topic deals with what we call "scope" in programming. When Python is looking for some variable, it first looks for it in the local scope, then the enclosing, then the global scope and finally the built-in scope (leading to the LEGB acronym). Scoping is out of the scope of this notebook (pun intended), but suffice it to say that in general we should alias functions to avoid name conflicts.

In [20]:
from numpy import mean as average
average([4, 8, 3])

5.0

So this is it with our short Python tutorial, but before we finish, let's give you some motivation for everything we learned.

At this point you might be wondering, am I here to learn data science or am I here to learn solve little programming challenges. In other words, how relevant is all of this to doing data science? The answer is very, very, very relevant, because knowing the basics well can help you write clear and concise code to manipulate data or train models. We'll look at some examples here, but the truth is it takes time and practice to come to this realization.

## Reading semi-structured data

Let's show an example of how the things we learned can be applied to a data science situation. We will go and read in some data from a JSON file, which is an example of what we call **semi-structured data**. 

Before reading the data, go and open it in an editor (the file name is `books.json`). A JSON file is not a Python object, but does it look **similar to** any of the Python objects we've encountered so far?

Let's now go and read the file into Python. After reading it, we will print the first element of it, to see what kind of data is there. When you try to print objects that can have nested information, it's helpful to pretty-print it, using the `pprint` library so the information is more presentable.

In [21]:
import json
with open('../Data/books.json',encoding = 'utf-8') as f:
    books_dict = json.load(f)

from pprint import pprint
pprint(books_dict[0]) # print information for the first book

{'_id': 1,
 'authors': ['W. Frank Ableson', 'Charlie Collins', 'Robi Sen'],
 'categories': ['Open Source', 'Mobile'],
 'isbn': '1933988673',
 'longDescription': 'Android is an open source mobile phone platform based on '
                    'the Linux operating system and developed by the Open '
                    'Handset Alliance, a consortium of over 30 hardware, '
                    'software and telecom companies that focus on open '
                    'standards for mobile devices. Led by search giant, '
                    'Google, Android is designed to deliver a better and more '
                    'open and cost effective mobile experience.    Unlocking '
                    "Android: A Developer's Guide provides concise, hands-on "
                    'instruction for the Android operating system and '
                    'development tools. This book teaches important '
                    'architectural concepts in a straightforward writing style '
                    

We are now going to extract particular pieces of information from the first book: the title, author, category, and ISBN of the book. However, we run into a problem: books can have multiple authors and multiple categories, and for reasons that will become clear soon we don't want to allow that. Instead we'll do this:

- When there are multiple authors, we will replicate the information once for each author.
- When there are multiple categories, we will just take the first one and ignore the rest.

In [22]:
for key in books_dict:
    print()

In [23]:
elem = books_dict[0] # pull out the first element
num_authors = len(elem['authors'])
print("The first book has {} authors.".format(num_authors))

The first book has 3 authors.


Also for reasons that will become clear soon, we create a **tuple** for storing the information, so here because we have three authors, we create three tuples called `row_1`, `row_2` and `row_3`.

In [24]:
row_1 = (elem['title'], 
         elem['authors'][0], # first author
         elem['categories'][0], # first category
         elem['isbn'])

row_2 = (elem['title'], 
         elem['authors'][1], # second author
         elem['categories'][0], # first category
         elem['isbn'])

row_3 = (elem['title'], 
         elem['authors'][2], # third author
         elem['categories'][0], # first category
         elem['isbn'])

In [25]:
print(row_1, row_2, row_3)

('Unlocking Android', 'W. Frank Ableson', 'Open Source', '1933988673') ('Unlocking Android', 'Charlie Collins', 'Open Source', '1933988673') ('Unlocking Android', 'Robi Sen', 'Open Source', '1933988673')


Now it's time to take the content from above and place it in a tabular data format. We will use the `sqlite3` library, which gives us access to a light-weight SQL database in Python. Note that we are only doing this to illustrate how data can flow from one format into another. Unless we insist on using SQL databases we learn later that there are other options. 

We don't actually need to have SQLite installed on our machine. Instead we use `sqlite.connect(':memory:')` to connect to a "database" in the memory and pretend it's a physical database somewhere.

In [26]:
import sqlite3

connection = sqlite3.connect(':memory:') 
cursor = connection.cursor()

cursor.execute('''CREATE TABLE books_long
             (title text, author text, categroy text, isbn text)''')

rows = [row_1, row_2, row_3]
cursor.executemany('INSERT INTO books_long VALUES (?,?,?,?)', rows)

connection.commit() # save the changes

Notice how in the above snippet, we first create a SQL table with column names and column types matching what we extracted from the JSON file. We then grouped `row_1`, `row_2` and `row_3` into a list and then inserted them into a SQL table we created, by using `INSERT INTO`.

### Exercise (3 minutes)

Based on what we learned about lists and tuples, see if you can answer the following questions:

1. What is the type of `row_1`? Provide a justification for this chioce.
1. What is the type of `row`? Provide a justification for this chioce.

### End of exercise

How do we check that it all worked? We can simply run a `SELECT *` on the data, and use `fetchall()` to grab it from the database and bring it back into Python.

In [27]:
books_table = cursor.execute("SELECT * FROM books_long").fetchall()
books_table

[('Unlocking Android', 'W. Frank Ableson', 'Open Source', '1933988673'),
 ('Unlocking Android', 'Charlie Collins', 'Open Source', '1933988673'),
 ('Unlocking Android', 'Robi Sen', 'Open Source', '1933988673')]

So let's summarize what we accomplished:

1. We read data from a JSON file into a Python dictionary.
1. We extracted some of the data out of the Python dictionary and placed it into a list of tuples.
1. We took the list of tuples and dumped its content into a SQL table.
1. We read the content back from the SQL table and into Python as a list of tuples.

As data flows from one format to another, it's important to think about the right object for representing the data. The right choice depends on many factors, such as

- The size of the data and other efficiency factors
- Whether the data is flat or hierarchical
- Whether the data is structured, unstructured, or semi-structured
- Our preference for how to "query" the data

Working with data, especially at scale, is such an important topic in data science that a relative new role, called **data engineer**, was created to deal with this. Among other things, data engineers work on creating efficient data pipelines to reduce redundancy, and letting the right tools do the job.

# Assignment

In this assignment we want to get comfortable with loading and manipulating data in Python. While future assignments will focus more using structured data which we can load into a `DataFrame` using `pandas`, this assignment is focused on semi-structured data and how we can "flatten" it and then load it into other formats. The objective is to see how data flows in Python from one object to another and what advantages and disadvantages each offers.

Let's read the `books.json` data set and display the first item in it.

In [28]:
import json
with open('../Data/books.json', encoding = 'utf-8') as f:
    books_dict = json.load(f)

1. Write a program that goes through the entire data and extracts the following information:  <span style="color:red" float:right>[4 point]</span>

  - title of the book
  - name of the first author
  - name of the second author (if book has more than one author)
  - number of authors
  - ISBN
  - if the word "data" is in the book's description
  - the number of words in the book's description
  - the year the book was published

  Of course because JSON data doesn't necessarily enforce any sort of schema, we can't be sure that the information we are trying to extract exists for every book. For example, if the book only has one author, then there is no second author. So use `try` and `except` as you loop through every book and skip to the next item every time some information is missing.

  Store the extracted data in a list named `rows` whose elements are tuples, one tuple per book. For example, the first element of `rows` stores the tuple for the first book and should look like this:

        ('Unlocking Android', 'W. Frank Ableson', 'Charlie Collins', 3, '1933988673', True, 252, 2009)

Reasoning - I struggled with this one, had a tough time getting the code right to include the correct amount of authors and getting it to loop correctly, Emailed about this one and did not hear back on it. Curious if there will be an option for potential zoom 'office hours' just to chat through some of these. I dont mind Piazza but asking a question and hearing back a few days later can be a bit frustrating rather than talking through it. 

Made an empty list called rows

Tried to loop through by using the elem and going through the entirity of the books_dict and extracting the 

In [29]:
rows = []    


In [30]:
elem = books_dict[1:432] # pull out the first element
##Testing out elem
elem[1]['title']

'Specification by Example'

In [31]:
for values in books_dict:
    rows+=(elem['title'], #Title of the book
         elem['authors'], #Name of first author
         elem['authors'], #Name of second author
         len(['authors']), #Number of authors
         elem['isbn'], #ISBN
         'data' in 'longDescription')
    

TypeError: list indices must be integers or slices, not str

In [ ]:
type(rows)

Conclusion - Was the data not stored correctly? When I tried to single out the first row it only gave me the authors

2. Save the content of `rows` in a SQL-like table using `sqlite3`, and choose the appropriate column types. <span style="color:red" float:right>[2 point]</span> 

  As your column names use the following:

  - `title`
  - `author_1`
  - `author_2`
  - `num_authors`
  - `isbn`
  - `has_data`
  - `desc_len`
  - `year_published`

Reasoning - imported sqlite3 and then chose the appropriate column types (books_long - think it got a little messed up due to rows not being properly stored

imported sqlite, added a connection to the rows list

executed a new table called books_long that included the title, author_1, author_2, num_authors, isbn, has_data, desc_len and year_published

In [ ]:
import sqlite3

connection = sqlite3.connect('rows') 
cursor = connection.cursor()

cursor.execute('''CREATE TABLE books_long_3
             (title, author_1, author_2, num_authors, isbn, has_data, desc_len, year_published)''')

rows = [row_1, row_2, row_3]
cursor.executemany('INSERT INTO rows VALUES (title, author_1, author_2, num_authors, isbn, has_data, desc_len, year_published)', books_long_3)

connection.commit() # save the changes

Conclusion - Another one I needed some help on walking through - mainly on the cursor execute many, thought we wanted the table ('books_long_3' in the final slot for the data???

3. Write a SQL query against the table to show all books that (1) contain the word "data" and (2) have more than 3 authors. Store the result of the query in an object called `books_table`, then close the connection. <span style="color:red" float:right>[2 point]</span>

Reasoning - working with fictional execute due to my code above not working 

Used the cursor.execute method while creating a new object called books_table

In [ ]:
books_table = cursor.execute('SELECT has_data, num_authors FROM books_long_3 WHERE has_data = TRUE AND num_authors > 3')

print(books_table)

Conclusion - came up with an empty query due to the above issues, I hope that the code is at least correct for this sequence to select the books with data and more than 3 authors

SQL tables are not the only way, and definitely not the most straightforward way to store and manipulate data in Python. A format that's more popular with data scientist is to use the `pandas` library to create a `DataFrame`. This library has a lot of functionality that makes it easy to run the common tasks data scientists do with data.

4. Read the data from the above query into a `DataFrame` and call it `books_df`. HINT: Use `pd.DataFrame` and specify the names to use for the columns. <span style="color:red" float:right>[1 point]</span>

Reasoning - Made up a dataframe of books from the lord of the rings series, did not want to keep running cells that did not work

imported pandas as pd

created a dataframe using the pd.DataFrame command and inserted books, author, and publishdate as columns

renamed the columns after as well

In [ ]:
###If the data was populated above then I would use the line in this cell
import pandas as pd

###Can specify the names used
books_df = pd.read_sql_query("SELECT * FROM has_data, num_authors", books_table)


In [ ]:
import pandas as pd

books_df = pd.DataFrame({'Books':['The Hobbit', 'The Return of the King', 'The Fellowship of the Ring', 'The Two Towers', 'A Promised Land', 'Hidden Valley Road'], 'Author':['J.R.R. Tolkein', 'J.R.R. Tolkein','J.R.R. Tolkein','J.R.R. Tolkein', 'Obama', 'Hello Name'],
                        'PublishDate':['1989', '1994', '1998', '2001', '2020', '2021'], 'len_authors':['1','4','7','2','1','3']})

In [ ]:
books_df

In [ ]:
books_df.rename(columns={'Author':'AUTHOR', 'PublishDate':'Year'}, inplace=True)

books_df

Conclusion - Had to create a fake dataset to work with due to issues above, was able to create a dataframe and then manipulate the columns to work with the questions posed

5. Display the first few columns of a `DataFrame` by calling its `head` method. <span style="color:red" float:right>[1 point]</span>

Reasoning - pretty straightforward, just called the head() and showed the first five rows of the data, can also use (X) if you dont want to display the top 5

In [ ]:
books_df.head()

In [ ]:
books_df.head(1)

Conclusion - Nice and easy way to visualize the data and to see what the columns and data is for a small snippet

Remember how earlier we said that a `DataFrame` is built on top of `numpy` arrays? Another way of saying it is that a `DataFrame` is an **abstraction** on top of `numpy` arrays: i.e. a `DataFrame` is a more **high-level** object than a `numpy` array. 

6. Call the `values` attribute of your `DataFrame` to convert it into a numpy array and display the first 3 elements of the array. <span style="color:red" float:right>[1 point]</span>

Reasoning - Had to importy the numpy libary as np first and then used the values attribute of the DataFrame to convert it into a numpy array (books) and then displaced the first 3 elemens

In [ ]:
import numpy as np

books = books_df.values

books[:4]

In [ ]:
###I am used to DataFrame.to_numpy

books_df.to_numpy()



Conclusion - Enjoy working with pandas and feel much more comfortable with this, sorry for the messy issues with question 1 + 2 - would you by any chance be willing to have a zoom session here or there to help me talk through some of the questions I have issues with?

Now you can judge which object is more "user-friendly". That's one of the things that abstractions allow us to do: build more user-friendly (abstract) objects from less user-friendly (but more fundamental) objects.

# End of assignment